staging/src/k8s.io/kube-aggregator/pkg/apiserver/retry_detector.go

```diff
@@ -0,0 +1,268 @@
+
+// newHijackErrorResponder creates a new ErrorResponder that wraps the delegate for supporting reties
+func newHijackErrorResponder(delegate proxy.ErrorResponder, req *http.Request) *retriableHijackErrorResponder {
+	return &retriableHijackErrorResponder{delegate: delegate, req: req}
+}
+
+// Error reports the err to the client or suppress it if it's not retriable
+func (hr *retriableHijackErrorResponder) Error(w http.ResponseWriter, r *http.Request, err error) {
+	// if we can retry the request do not send a response to the client
+	hr.lastKnownError = err
+	hr.req = r
+	if !hr.isRetriable(hr.req, err) {
+		hr.delegate.Error(w, r, err) // this might send a response to a client
+	}
+}
+
+func (hr *retriableHijackErrorResponder) reset() {
+	hr.lastKnownError = nil
+}
+
+func (hr *retriableHijackErrorResponder) retry() bool {
+	return hr.isRetriable(hr.req, hr.lastKnownError)
+}
+
+// withHijackErrorResponderForSingleEndpoint is used by the ErrorResponder to determine
+// if the given error along with the request is safe to retry when only single endpoint is available
+func withHijackErrorResponderForSingleEndpoint(hr *retriableHijackErrorResponder) retriable {
+	hr.isRetriable = func(req *http.Request, err error) bool {
+		if isHTTPVerbRetriable(req) && (knet.IsConnectionReset(err) || knet.IsProbableEOF(err) || isExperimental(err)) {
+			return true
+		}
+		return false
+	}
+	return hr
+}
+
+// withHijackErrorResponderForMultipleEndpoints is used by the ErrorResponder to determine
+// if the given error along with the request is safe to retry when more than one endpoint is available
+func withHijackErrorResponderForMultipleEndpoints(hr *retriableHijackErrorResponder) retriable {
+	hr.isRetriable = func(req *http.Request, err error) bool {
+		// we always want to retry connection refused errors as the aggregator will pick up a different host on the next try.
+		// the error is of particular interest during graceful shutdown of the backend server:
+		// 	net/http2 library automatically retries requests/streams after receiving a GOAWAY frame.
+		// 	this is true for StreamsWithID > LastStreamID (present in the GOAWAY frame) which represents the last stream identifier the server has processed or is aware of.
+		// 	StreamsWithID <= LastStreamID might be fully processed because the server waits until all current streams are done or the timeout expires.
+		// 	in case of the timeout, we will get "http2: server sent GOAWAY and closed the connection" error which is handled by IsProbableEOF() and it is safe to retry only for particular verbs (check isHTTPVerbRetriable method)
+		// 	on the next try a new connection will be opened to the same host and will fail with "connection refused" error because the remote host stopped listening on the port.
+		if knet.IsConnectionRefused(err) {
+			return true
+		}
+		if isHTTPVerbRetriable(req) && (knet.IsConnectionReset(err) || knet.IsProbableEOF(err) || isExperimental(err)) {
```

Comment:
I understand that some verbs can be retried and I understand why.  However, this particular one cheats the test you're trying to fix.  I can update the test to make sure it doesn't trip this, but I'd rather limit the branches in our proxy overall.